keanu-project/src/main/java/io/improbable/keanu/algorithms/mcmc/Hamiltonian.java

```diff
@@ -96,15 +118,72 @@ public NetworkSamples getPosteriorSamples(final BayesianNetwork bayesNet,
 
         final Map<VertexId, ?> sampleBeforeLeapfrog = new HashMap<>();
 
-        for (int sampleNum = 1; sampleNum < sampleCount; sampleNum++) {
+        return new HamiltonianSampler(latentVertices, position, gradient, momentum, positionBeforeLeapfrog,
+            gradientBeforeLeapfrog, momentumBeforeLeapfrog, random, fromVertices, leapFrogCount,
+            stepSize, logProbGradientCalculator, sampleBeforeLeapfrog, bayesNet, logOfMasterPBeforeLeapfrog
+        );
+    }
+
+    public static class HamiltonianSampler implements SamplingAlgorithm {
+
+        private List<Vertex<DoubleTensor>> latentVertices;
+        private Map<VertexId, DoubleTensor> position;
+        private Map<VertexId, DoubleTensor> gradient;
+        private Map<VertexId, DoubleTensor> momentum;
+        private Map<VertexId, DoubleTensor> positionBeforeLeapfrog;
+        private Map<VertexId, DoubleTensor> gradientBeforeLeapfrog;
+        private Map<VertexId, DoubleTensor> momentumBeforeLeapfrog;
+        private KeanuRandom random;
+        private List<? extends Vertex> fromVertices;
+        private int leapFrogCount;
+        private double stepSize;
+        private LogProbGradientCalculator logProbGradientCalculator;
+        private Map<VertexId, ?> sampleBeforeLeapfrog;
+        private BayesianNetwork bayesNet;
+        private double logOfMasterPBeforeLeapfrog;
+
+        public HamiltonianSampler(List<Vertex<DoubleTensor>> latentVertices,
+                       Map<VertexId, DoubleTensor> position,
+                       Map<VertexId, DoubleTensor> gradient,
+                       Map<VertexId, DoubleTensor> momentum,
+                       Map<VertexId, DoubleTensor> positionBeforeLeapfrog,
+                       Map<VertexId, DoubleTensor> momentumBeforeLeapfrog,
+                       Map<VertexId, DoubleTensor> gradientBeforeLeapfrog,
+                       KeanuRandom random,
+                       List<? extends Vertex> fromVertices,
+                       int leapFrogCount,
+                       double stepSize,
+                       LogProbGradientCalculator logProbGradientCalculator,
+                       Map<VertexId, ?> sampleBeforeLeapfrog,
+                       BayesianNetwork bayesNet,
+                       double logOfMasterPBeforeLeapfrog) {
+
+            this.latentVertices = latentVertices;
+            this.position = position;
+            this.gradient = gradient;
+            this.momentum = momentum;
+            this.positionBeforeLeapfrog = positionBeforeLeapfrog;
+            this.momentumBeforeLeapfrog = momentumBeforeLeapfrog;
+            this.gradientBeforeLeapfrog = gradientBeforeLeapfrog;
```

Comment:
it looks like (position, gradient, momentum) belong together - put them into a new class?